From a69ec8ca7464ad10c55eafbcda68ccac37d3c714 Mon Sep 17 00:00:00 2001
From: Josua Mayer <josua@solid-run.com>
Date: Mon, 27 Nov 2023 14:30:52 +0100
Subject: [PATCH 2/8] arm64: dts: ti: k3-am64-main: fix pci-e gic interrupt
 numbers

Existing pci-e nodes refer to gic interrupts 202 and 203 for legacy
interrupts and link state.

According to TRM for AM64x/AM243x Processors, PCI-E Interrupts are:
- Legacy Interrupt: 234 (PCIE0_PCIE_LEGACY_PULSE_0)
- Link State: 225  (PCIE0_PCIE_LINK_STATE_PULSE_0)
Update the references accordingly.

This fixes various stalls encountered when bringing up wifi modules
(ifconfig up / iw dev wlan0 scan), such as e.g. AR9462:
[  112.659714] rcu: INFO: rcu_preempt self-detected stall on CPU
[  112.665480] rcu:     0-....: (1 GPs behind) idle=dd94/1/0x4000000000000000 softirq=1887/1888 fqs=2625
[  112.674430]  (t=5252 jiffies g=1113 q=16 ncpus=2)
[  112.679130] CPU: 0 PID: 40 Comm: kworker/u4:1 Not tainted 6.1.46-00001-gdd63f4a25737-dirty #16
[  112.687727] Hardware name: SolidRun AM642 HummingBoard-T with PCI-E (DT)
[  112.694414] Workqueue: phy0 ieee80211_scan_work [mac80211]
[  112.700081] pstate: 80000005 (Nzcv daif -PAN -UAO -TCO -DIT -SSBS BTYPE=--)
[  112.707030] pc : ath9k_ioread32+0x30/0xa4 [ath9k]
[  112.711756] lr : __ath9k_hw_enable_interrupts+0x114/0x2c0 [ath9k_hw]
[  112.718185] sp : ffff80000a7fbaa0
[  112.721489] x29: ffff80000a7fbaa0 x28: ffff0000060788e0 x27: ffff0000060788e0
[  112.728618] x26: ffff00000607bf70 x25: ffff000006f04080 x24: ffff00000607ab9c
[  112.735746] x23: ffff00000607c070 x22: ffff00000607c818 x21: 0000000000000002
[  112.742874] x20: 00000000000000a0 x19: 0000000081800175 x18: ffff000006f04957
[  112.750002] x17: ffff000006f047ea x16: 00000000000012c0 x15: 0000000000000004
[  112.757129] x14: 0000000000000000 x13: 0000000000000004 x12: 0000000000000007
[  112.764256] x11: 0000000000000003 x10: ffff000006078a48 x9 : 0000000000000600
[  112.771383] x8 : 0000000000000280 x7 : 0000000000000400 x6 : 0000000000023f40
[  112.778510] x5 : 00000000f4000473 x4 : 000000000000401c x3 : ffff800001278fd0
[  112.785637] x2 : ffff800001279710 x1 : 0000000000000000 x0 : ffff000006f04080
[  112.792765] Call trace:
[  112.795204]  ath9k_ioread32+0x30/0xa4 [ath9k]
[  112.799575]  __ath9k_hw_enable_interrupts+0x114/0x2c0 [ath9k_hw]
[  112.805633]  ath9k_hw_enable_interrupts.part.0+0x50/0x60 [ath9k_hw]
[  112.811951]  ath9k_hw_enable_interrupts+0x18/0x30 [ath9k_hw]
[  112.817661]  ath_complete_reset+0x84/0x190 [ath9k]
[  112.822464]  ath_reset_internal+0x11c/0x290 [ath9k]
[  112.827353]  ath_reset+0x4c/0x90 [ath9k]
[  112.831289]  ath_set_channel.isra.0+0x104/0x1c0 [ath9k]
[  112.836525]  ath_chanctx_set_channel+0x74/0x90 [ath9k]
[  112.841675]  ath9k_config+0x1a0/0x1f0 [ath9k]
[  112.846043]  ieee80211_hw_config+0x88/0x310 [mac80211]
[  112.851291]  ieee80211_scan_work+0x468/0x6a4 [mac80211]
[  112.856620]  process_one_work+0x1d0/0x320
[  112.860630]  worker_thread+0x14c/0x444
[  112.864373]  kthread+0x10c/0x110
[  112.867596]  ret_from_fork+0x10/0x20

Fixes: ac36a376e ("arm64: dts: ti: k3-am64-main: Add properties to support legacy interrupts")
Signed-off-by: Josua Mayer <josua@solid-run.com>
---
 arch/arm64/boot/dts/ti/k3-am64-main.dtsi | 6 +++---
 1 file changed, 3 insertions(+), 3 deletions(-)

diff --git a/arch/arm64/boot/dts/ti/k3-am64-main.dtsi b/arch/arm64/boot/dts/ti/k3-am64-main.dtsi
index 7b4cc0578d20..7bd862d3abeb 100644
--- a/arch/arm64/boot/dts/ti/k3-am64-main.dtsi
+++ b/arch/arm64/boot/dts/ti/k3-am64-main.dtsi
@@ -861,7 +861,7 @@ pcie0_rc: pcie@f102000 {
 		      <0x00 0x68000000 0x00 0x00001000>;
 		reg-names = "intd_cfg", "user_cfg", "reg", "cfg";
 		interrupt-names = "link_state";
-		interrupts = <GIC_SPI 203 IRQ_TYPE_EDGE_RISING>;
+		interrupts = <GIC_SPI 235 IRQ_TYPE_EDGE_RISING>;
 		device_type = "pci";
 		ti,syscon-pcie-ctrl = <&main_conf 0x4070>;
 		max-link-speed = <2>;
@@ -891,7 +891,7 @@ pcie0_intc: interrupt-controller {
 			interrupt-controller;
 			#interrupt-cells = <1>;
 			interrupt-parent = <&gic500>;
-			interrupts = <GIC_SPI 202 IRQ_TYPE_EDGE_RISING>;
+			interrupts = <GIC_SPI 234 IRQ_TYPE_EDGE_RISING>;
 		};
 	};
 
@@ -903,7 +903,7 @@ pcie0_ep: pcie-ep@f102000 {
 		      <0x00 0x68000000 0x00 0x08000000>;
 		reg-names = "intd_cfg", "user_cfg", "reg", "mem";
 		interrupt-names = "link_state";
-		interrupts = <GIC_SPI 203 IRQ_TYPE_EDGE_RISING>;
+		interrupts = <GIC_SPI 235 IRQ_TYPE_EDGE_RISING>;
 		ti,syscon-pcie-ctrl = <&main_conf 0x4070>;
 		max-link-speed = <2>;
 		num-lanes = <1>;
-- 
2.35.3

